improve test cleanup by removing any docker containers based on test image

mmathesius opened the pull request:
I discovered the setup/teardown scripts didn't check for any lingering containers based on the base-runtime-smoke image before attempting to remove it. If there were any, the cleanup would fail and require manual interventions. This update fixes that problem.

mmathesius committed:
improve cleanup by removing any docker containers based on test image

mmathesius committed:
correct exception handling

sgallagher reviewed (changes requested):
This is an awful lot of code duplication; could you just turn this into a common function that is called in both locations, please?

mmathesius committed:
reduce code duplication by moving docker container/image and mock root cleanup
to common function used by both setup and teardown tests

mmathesius commented:
I disliked the redundancy myself, but I wasn't sure of the right way to go about it with Avocado. But I've figured it out now. (It was mostly a matter of getting a handle on the Avocado logger...)

sgallagher commented on cleanup.py:
```diff
@@ -0,0 +1,71 @@
+                stderr = subprocess.STDOUT, shell = True)
+        except subprocess.CalledProcessError as e:
+            log.error("command '%s' returned exit status %d; output:\n%s" %
+                (e.cmd, e.returncode, e.output))
```
If we fail here, we should probably raise an exception that we can catch in the tests so we don't attempt to continue (since the environment isn't clean).

sgallagher commented on cleanup.py:
```diff
@@ -0,0 +1,71 @@
+    except subprocess.CalledProcessError as e:
+        if "No such image" not in e.output:
+            log.error("command '%s' returned exit status %d; output:\n%s" %
+                (e.cmd, e.returncode, e.output))
```
Same here; if it fails, the rest of the tests will be non-deterministic.

sgallagher commented on cleanup.py:
```diff
@@ -0,0 +1,71 @@
+            stderr = subprocess.STDOUT)
+    except subprocess.CalledProcessError as e:
+        log.error("command '%s' returned exit status %d; output:\n%s" %
+            (e.cmd, e.returncode, e.output))
```
Ditto

sgallagher commented on setup.py:
```diff
@@ -40,31 +42,8 @@ def setUp(self):
-        self.log.info("mock teardown with '%s' succeeded with output:\n%s" %
-            (mock_teardown_cmdline, mock_teardown_output))
+        # Clean-up any old test artifacts (docker containers, image, mock root) first:
+        cleanup.cleanup_docker_and_mock(self.mockcfg)
```
Wrap this in a try/except and handle cleanup failures gracefully.

sgallagher commented on teardown.py:
```diff
@@ -29,33 +31,7 @@ def setUp(self):
-
-
-
+        cleanup.cleanup_docker_and_mock(self.mockcfg)
```
Wrap this in a try/except and handle cleanup failures gracefully.

mmathesius committed:
Improve exception handling during artifact cleanup to prevent attempts to
continue after a fatal error. It is now necessary to re-raise exceptions
when they are caught during cleanup since Avocado can't do it automatically
when errors are logged in an external function.

mmathesius commented:
Excellent points. I overlooked the fact that Avocado was no longer automatically raising exceptions when an error was logged by the (now external) cleanup routine.
